File: crates/idakit-sys-codegen/src/domains/function.rs
use super::super::model::*;

/// The function domain: per-function scalar accessors and the name string. Function *chunks* are
/// the `range` domain (`range_all_chunks`), so no chunk accessor lives here. `func_qty` is a
/// templated passthrough; the lookup accessors are hand-written in `facade/function.cpp`.
pub const FUNCTION: Domain = Domain {
    name: "function",
    // compat.h supplies func_entry_info_t and the GFI_* selector under 9.3.
    sdk_includes: &["<funcs.hpp>", "<name.hpp>", "<stdexcept>", "\"compat.h\""],
    externs: &[],
    structs: &[SharedStruct {
        name: "FunctionEntryInfo",
        doc: "One function entry chunk's scalars, plus the strings the caller asked for, returned \
              by value from [`func_entry_info`]. An unrequested or absent string is empty.",
        fields: fields! {
            start: U64 = "Entry address.";
            end: U64 = "Entry-chunk end address, exclusive.";
            flags: U64 = "Raw `func_t::flags` bits (`FUNC_*`).";
            frame_id: U64 = "Netnode id of the frame structure, `BADNODE` when there is none.";
            locals_size: U64 = "Size in bytes of the frame's local-variable part (`frsize`).";
            saved_regs_size: U32 = "Size in bytes of the saved-registers part of the frame.";
            purged_bytes: U64 = "Bytes purged from the stack on return.";
            frame_pointer_delta: U64 = "Frame pointer delta, usually 0.";
            color: U32 = "User-defined function color, `DEFCOLOR` when unset.";
            name: Str = "Function name, empty when not requested.";
            cmt: Str = "Regular comment, empty when not requested or absent.";
            cmt_rpt: Str = "Repeatable comment, empty when not requested or absent.";
        },
    }],
    consts: &[],
    custom_tus: &["facade/function.cpp"],
    fns: fns! {
        "Number of functions in the database (`get_func_qty`)."
            func_qty() -> Usize = scalar("get_func_qty()");
        "Entry address of the `n`-th function, or `BADADDR` when `n` is out of range."
            func_ea(n: Usize) -> U64;
        "Entry address of the function containing `ea`, or `BADADDR` when there is none."
            func_start(ea: U64) -> U64;
        "Entry-chunk end address of the function at `ea`, or `BADADDR` when not a function."
            func_end(ea: U64) -> U64;
        "`func_t::flags` of the function at `ea`, or `0` when `ea` is not a function."
            func_flags(ea: U64) -> U64;
        "Number of chunks (entry plus tails) of the function at `ea`, or `0`."
            func_chunk_qty(ea: U64) -> I32;
        "Name of the function at `ea`; `Err` when it has none."
            func_name(ea: U64) -> ResultString;
        "Comment of the function at `ea` (repeatable if `repeatable`); `Err` when there is none."
            func_cmt(ea: U64, repeatable: Bool) -> ResultString;
        "Whether the function at `callee` is known to return (`func_does_return`)."
            func_does_return(callee: U64) -> Bool = scalar("::func_does_return(static_cast<ea_t>(callee))");
        "Addressing width in bits of the function at `ea`: 16, 32, or 64, or 0 when not a function."
            func_bitness(ea: U64) -> I32;
        "Entry-chunk scalars of the function at `ea`, plus each string `fields` selects (a \
         `GFI_*` mask); `Err` when `ea` is not a function."
            func_entry_info(ea: U64, fields: I32) -> ResultShared("FunctionEntryInfo");
    },
};

File: crates/idakit-sys/facade/function.cpp
// Hand-written Custom bodies for the generated function domain (namespace gen): scalar lookup
// accessors keyed by address returning BADADDR (or 0) when there is no function, the name and
// comment as rust::String (throw, a Rust Err, when absent), and the function's bitness.
// func_qty and func_does_return are templated (gen_function_bodies.cc), not here.

#include <ida.hpp>
#include <pro.h>

#include <funcs.hpp>
#include <name.hpp>

#include <stdexcept>

#include "gen_function.h"
// The generated bridge header defines FunctionEntryInfo (needed in full to construct it below);
// gen_function.h only forward-declares it.
#include "gen_bridge.h"

#include "compat.h"

namespace gen {

// Start address of the nth function in the functions list, or BADADDR when n is out of range.
uint64_t func_ea(size_t n) { return static_cast<uint64_t>(get_func_ea_by_num(n)); }

// Start address of the function containing addr, or BADADDR when there is none.
uint64_t func_start(uint64_t addr) {
  return static_cast<uint64_t>(get_func_start(static_cast<ea_t>(addr)));
}

// End address (exclusive) of the function containing addr, or BADADDR when there is none. Keyed on
// the entry chunk, so an address in a tail answers for the whole function.
uint64_t func_end(uint64_t addr) {
  fchunk_info_t entry;
  if (!get_fchunk_info(&entry, get_func_start(static_cast<ea_t>(addr))))
    return static_cast<uint64_t>(BADADDR);
  return static_cast<uint64_t>(entry.end_ea);
}

// The function's flags word at addr, or 0 when addr is not inside a function.
uint64_t func_flags(uint64_t addr) {
  fchunk_info_t entry;
  if (!get_fchunk_info(&entry, get_func_start(static_cast<ea_t>(addr))))
    return 0;
  return entry.get_flags();
}

// Number of chunks (main body plus tails) making up the function at addr, or 0 when there is none.
int32_t func_chunk_qty(uint64_t addr) {
  ea_t func_ea = get_func_start(static_cast<ea_t>(addr));
  if (func_ea == BADADDR)
    return 0;
  return static_cast<int32_t>(get_func_tail_qty(func_ea)) + 1;
}

// The function's name at addr; throws when the function is missing or unnamed.
rust::String func_name(uint64_t addr) {
  qstring out;
  if (get_func_name(&out, static_cast<ea_t>(addr)) <= 0)
    throw std::runtime_error("no function name at address");
  return to_rust_string(out);
}

// The function's comment at addr (repeatable or regular); throws when addr is not a function or
// that channel carries no comment.
rust::String func_cmt(uint64_t addr, bool repeatable) {
  ea_t ea = static_cast<ea_t>(addr);
  if (get_func_start(ea) == BADADDR)
    throw std::runtime_error("no function at address");
  qstring out;
  if (get_func_cmt_ea(&out, ea, repeatable) <= 0)
    throw std::runtime_error("no function comment");
  return to_rust_string(out);
}

// The function's addressing width in bits at addr: 16, 32, or 64, or 0 when addr is not a
// function. Returns the width (get_func_bits_ea), not the raw 0/1/2 bitness code, so the Rust
// side's width-based Bitness conversion applies uniformly across every bitness accessor.
int32_t func_bitness(uint64_t addr) {
  ea_t ea = static_cast<ea_t>(addr);
  return get_func_start(ea) != BADADDR ? static_cast<int32_t>(get_func_bits_ea(ea)) : 0;
}

// The entry chunk's scalars in one call, plus each string the GFI_* mask selects; throws when addr
// is not a function. An unselected string comes back empty rather than costing a build.
FunctionEntryInfo func_entry_info(uint64_t addr, int32_t fields) {
  func_entry_info_t entry;
  if (!get_func_entry_info(&entry, static_cast<ea_t>(addr), fields))
    throw std::runtime_error("no function at address");
  FunctionEntryInfo out;
  out.start = static_cast<uint64_t>(entry.start_ea);
  out.end = static_cast<uint64_t>(entry.end_ea);
  out.flags = entry.get_flags();
  out.frame_id = static_cast<uint64_t>(entry.get_frame_id());
  out.locals_size = static_cast<uint64_t>(entry.get_frsize());
  out.saved_regs_size = static_cast<uint32_t>(entry.get_frregs());
  out.purged_bytes = static_cast<uint64_t>(entry.get_argsize());
  out.frame_pointer_delta = static_cast<uint64_t>(entry.get_fpd());
  out.color = static_cast<uint32_t>(entry.get_color());
  out.name = rust::String::lossy(entry.get_name());
  out.cmt = rust::String::lossy(entry.get_cmt());
  out.cmt_rpt = rust::String::lossy(entry.get_cmt_rpt());
  return out;
}

} // namespace gen
